lwip_demo/lwip_demo.c

```diff
@@ -0,0 +1,333 @@
+
+   SMEMCPY(netif->hwaddr, MYMACADDRESS, sizeof(netif->hwaddr));
+   netif->hwaddr_len = sizeof(netif->hwaddr);
+   return ERR_OK;
+}
+
+void init()
+{
+
+}
+
+#if LWIP_DHCP
+/*
+ * DHCP Init
+ */
+void dhcp_init(struct netif *netif)
+{
+   err_t error_dhcp;
+   /* Start DHCP and HTTPD */
+   error_dhcp = dhcp_start(netif);
+   if (error_dhcp != ERR_OK) {
+      printf("DHCP Error occurred - out of memory.\n");
+   } else {
+      printf("DHCP started.\n");
+      u8_t myip_dhcp;
+      myip_dhcp = dhcp_supplied_address(netif);
+      printf("main: ip address now: %x\n", myip_dhcp);
+   }
+}
+#endif // LWIP_DHCP
+
+void main(void)
+{
+   app_init();
+   lwip_init();
+   struct netif netif;
+   struct netif *netif_control;
+   err_t err_ipv6_netif;
+
+   // IP Address for example packets
+   IP4_ADDR(&gw, 192, 168, 100, 1);
+   IP4_ADDR(&ipaddr, 192, 168, 100, 114);
+   IP6_ADDR(&ip6addr, 1234, 5666, 1914, 5111);
+   IP4_ADDR(&netmask, 255, 255, 255, 0);
+
+   netif_control = netif_add(&netif, &ipaddr, &netmask, &gw, NULL, my_init,
+                             netif_input);
+
+   if (netif_control == NULL) {
+      printf("netif_add_ip6_address failed.\n");
+   }
+
+   err_ipv6_netif = netif_add_ip6_address(&netif, &ip6addr, NULL);
+   if (err_ipv6_netif != ERR_OK) {
+      printf("netif_add_ip6_address failed.\n");
+   }
+
+   netif.name[0] = 'e';
+   netif.name[1] = '0';
+
+   netif_create_ip6_linklocal_address(&netif, 1);
+   netif.ip6_autoconfig_enabled = 1;
+
+   netif_set_status_callback(&netif, netif_status_callback);
+   netif_set_default(&netif);
+   netif_set_up(&netif);
+
+#if LWIP_DHCP
+   dhcp_init(&netif);
+#endif // LWIP_DHCP
+
+   // All initialization done, we're ready to receive data
+   printf("main: Reset done, Init done, Interrupts enabled\n");
+   printf("main: IER Register: %x\n", *IER);
+
+   optimsoc_list_iterator_t iter = 0;
+   eth_rx_pbuf_queue = optimsoc_list_init(NULL);
+   eth_rx_pbuf_queue = NULL;
+
+#if LWIP_UDP
+   static struct udp_pcb *udpecho_raw_pcb;
+   udp_init();
+   udpecho_raw_pcb = udp_new_ip_type(IPADDR_TYPE_ANY);
+   udp_recv(udpecho_raw_pcb, NULL, NULL);
+   udp_bind_netif(udpecho_raw_pcb, &netif);
+#endif // LWIP_UDP
+
+#if LWIP_DEBUG
+   debug_flags |= (LWIP_DBG_ON|LWIP_DBG_TRACE|LWIP_DBG_STATE|LWIP_DBG_FRESH|LWIP_DBG_HALT); // |LWIP_DBG_TRACE
+#endif //LWIP_DEBUG
+
+#if LWIP_TCP
+   static struct tcp_pcb *tcpecho_raw_pcb;
+   tcpecho_raw_pcb = tcp_new_ip_type(IPADDR_TYPE_ANY);
+   if (tcpecho_raw_pcb != NULL) {
+      tcp_bind(tcpecho_raw_pcb, IP_ANY_TYPE, 2049);
+      tcp_bind_netif(tcpecho_raw_pcb, &netif);
+   }
+#endif // LWIP_TCP
+
+   while (1) {
+      // TODO: Check link status
+
+      if (eth_rx_pbuf_queue != NULL
+            && optimsoc_list_length(eth_rx_pbuf_queue) != 0) {
+         if (NULL == optimsoc_list_first_element(eth_rx_pbuf_queue, &iter)) {
+            // printf("main: Element was NULL, return!\n");
+            eth_rx_pbuf_queue = NULL;
```

Comment:
This most likely is a memory leak, leaking the instance of optimsoc_list. Why do you set the list pointer to NULL? It's not problematic if the list doesn't have any entries. Anyways, this code should never be executed since you check if the list has entries in line 298 before.